### PR TITLE
Tvdb v4 migration

### LINKED ISSUE #47: Update for API v4 support
SDK: https://www.nuget.org/packages/Tvdb.Sdk
Deadline: ~Oct 1, 2021~ First half of 2022

### LINKED ISSUE #100: Legacy API keys are deprecated and will be discontinued soon. Developers should update their projects to make use of our new API.
I don't have a Legacy keys of TheTVDB, so i only have the API keys.

I got the error message when i use API keys.

```
TvDbSharper.TvDbServerException: Invalid credentials; API Key Required
   at TvDbSharper.Infrastructure.Parser.Parse[T](ApiResponse response, IReadOnlyDictionary`2 errorMap)
   at TvDbSharper.Clients.AuthenticationClient.AuthenticateAsync(AuthenticationData authenticationData, CancellationToken cancellationToken)
   at Jellyfin.Plugin.Tvdb.TvdbClientManager.GetTvDbClient(String language)
   at Jellyfin.Plugin.Tvdb.TvdbClientManager.<>c__DisplayClass24_0`1.<<TryGetValue>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.Extensions.Caching.Memory.CacheExtensions.GetOrCreateAsync[TItem](IMemoryCache cache, Object key, Func`2 factory)
   at Jellyfin.Plugin.Tvdb.TvdbClientManager.GetImageKeyTypesForSeriesAsync(Int32 tvdbId, String language, CancellationToken cancellationToken)+MoveNext()
   at Jellyfin.Plugin.Tvdb.TvdbClientManager.GetImageKeyTypesForSeriesAsync(Int32 tvdbId, String language, CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
   at Jellyfin.Plugin.Tvdb.Providers.TvdbSeriesImageProvider.GetImages(BaseItem item, CancellationToken cancellationToken)
   at Jellyfin.Plugin.Tvdb.Providers.TvdbSeriesImageProvider.GetImages(BaseItem item, CancellationToken cancellationToken)
   at MediaBrowser.Providers.Manager.ProviderManager.GetImages(BaseItem item, IRemoteImageProvider provider, String preferredLanguage, Boolean includeAllLanguages, CancellationToken cancellationToken, Nullable`1 type)
```

### PR DESCRIPTION
**Pull request by scampower3**
Tvdb v4 migation in progress. Will set on draft till its done. 
Some code commented out needs some clarification.

Fixes #47
Fixes #100 

~~Using this fork of TVdbSharper: https://github.com/scampower3/TvDbSharper~~
Switched to: https://github.com/crobibero/tvdb-sdk-csharp

### CODE REVIEW

**Comment by scampower3** [comment]
Should be working without any bugs now.

File: Jellyfin.Plugin.Tvdb/Configuration/PluginConfiguration.cs
Comment (crobibero) [inline Jellyfin.Plugin.Tvdb/Configuration/PluginConfiguration.cs:1]: This file's changes should be reverted, I will include the api key in a separate PR

File: Jellyfin.Plugin.Tvdb/TvdbUtils.cs
Comment (crobibero) [inline Jellyfin.Plugin.Tvdb/TvdbUtils.cs:64]: I'm pretty sure the `CultureInfo` object isn't populated when running in our docker container

Also, use `.LeftPart` if possible (part of Jellyfin.Extensions)

File: Jellyfin.Plugin.Tvdb/Providers/TvdbEpisodeProvider.cs
Comment (crobibero) [inline Jellyfin.Plugin.Tvdb/Providers/TvdbEpisodeProvider.cs:244]:
```suggestion
                var absoluteInfo = episode.Seasons.FirstOrDefault(x => string.Equals(x.Type.Name, "absolute", StringComparison.OrdinalIgnoreCase));
```

File: Jellyfin.Plugin.Tvdb/Providers/TvdbEpisodeProvider.cs
Comment (crobibero) [inline Jellyfin.Plugin.Tvdb/Providers/TvdbEpisodeProvider.cs:268]: same

File: Jellyfin.Plugin.Tvdb/Providers/TvdbEpisodeProvider.cs
Comment (crobibero) [inline Jellyfin.Plugin.Tvdb/Providers/TvdbEpisodeProvider.cs:277]: same

File: Jellyfin.Plugin.Tvdb/Providers/TvdbEpisodeProvider.cs
Comment (crobibero) [inline Jellyfin.Plugin.Tvdb/Providers/TvdbEpisodeProvider.cs:285]: same

File: Jellyfin.Plugin.Tvdb/Providers/TvdbEpisodeProvider.cs
Comment (crobibero) [inline Jellyfin.Plugin.Tvdb/Providers/TvdbEpisodeProvider.cs:293]: same

**Review by crobibero — COMMENTED** [review]
initial pass

File: Jellyfin.Plugin.Tvdb/Providers/TvdbEpisodeProvider.cs
Comment (scampower3) [reply Jellyfin.Plugin.Tvdb/Providers/TvdbEpisodeProvider.cs:215]: Hmm, there might be a chance that the english translation is not available. Maybe we can fall back to the series native language? Like Japanese show will fall back to Japanese

Comment (crobibero) [reply Jellyfin.Plugin.Tvdb/Providers/TvdbEpisodeProvider.cs:215]: Yeah that sounds more correct

File: Jellyfin.Plugin.Tvdb/TvdbUtils.cs
Comment (scampower3) [reply Jellyfin.Plugin.Tvdb/TvdbUtils.cs:64]: Alright. There is a way I can do a lookup for the languages that Jellyfin support?

File: Jellyfin.Plugin.Tvdb/Configuration/PluginConfiguration.cs
Comment (scampower3) [reply Jellyfin.Plugin.Tvdb/Configuration/PluginConfiguration.cs:1]: How about I change it to String.Empty and you update the API key accordingly?

File: Jellyfin.Plugin.Tvdb/Providers/TvdbEpisodeProvider.cs
Comment (scampower3) [reply Jellyfin.Plugin.Tvdb/Providers/TvdbEpisodeProvider.cs:222]: Yep, I'll change the == to string.Equals

File: Jellyfin.Plugin.Tvdb/TvdbUtils.cs
Comment (scampower3) [reply Jellyfin.Plugin.Tvdb/TvdbUtils.cs:65]: Hmm, is there a way for me to access Emby.Server.Implementations.Localization.iso6392.txt from the plugin?

**Comment by scampower3** [comment]
Should I include the TVSharper DLL inside the PR? Or is it better if you build the DLL from my [TvDbSharper](https://github.com/scampower3/TvDbSharper) fork?

File: Jellyfin.Plugin.Tvdb/TvdbUtils.cs
Comment (crobibero) [reply Jellyfin.Plugin.Tvdb/TvdbUtils.cs:65]: Sadly I don't think so, that project isn't meant to be referenced by plugins
I'm not sure how to correct this though

**Comment by crobibero** [comment]
> Should I include the TVSharper DLL inside the PR? Or is it better if you build the DLL from my [TvDbSharper](https://github.com/scampower3/TvDbSharper) fork?

I don't think so- I would say this PR is blocked until upstream merges your changes (https://github.com/HristoKolev/TvDbSharper/pull/39)

Alternatively we can migrate to the sdk I have generating (https://github.com/crobibero/tvdb-sdk-csharp)

**Comment by scampower3** [comment]
Hmm, I doubt he will merge my changes anytime soon since the last commit for that project is in 2022. Maybe we can keep this as an unoffical tvdb plugin until the changes are merged or we migrate to the sdk?

File: Jellyfin.Plugin.Tvdb/TvdbUtils.cs
Comment (scampower3) [reply Jellyfin.Plugin.Tvdb/TvdbUtils.cs:65]: Guess I can just dupe the txt file and build reference from there?

**Comment by luevano** [comment]
Hi @scampower3 any updates on this? wondering if this is still being worked on, or is it going to be fixed by another PR

**Comment by scampower3** [comment]
> Hi @scampower3 any updates on this? wondering if this is still being worked on, or is it going to be fixed by another PR

Hey, this is PR isnt abandoned. This migration is more or less done last I checked. However, it is dependent on the changes made [here](https://github.com/crobibero/tvdb-sdk-csharp/pull/22). This PR is probably blocked till the schema on TVDB site is fixed.

File: Jellyfin.Plugin.Tvdb/Providers/TvdbSeriesProvider.cs
Comment (scampower3) [reply Jellyfin.Plugin.Tvdb/Providers/TvdbSeriesProvider.cs]: Overlooked this part. At this commit [5ea6634](https://github.com/jellyfin/jellyfin-plugin-tvdb/pull/93/commits/5ea66341e1ce96011da32c4125ee197f7cd960eb)

File: Jellyfin.Plugin.Tvdb/Providers/TvdbMissingEpisodeProvider.cs
Comment (scampower3) [reply Jellyfin.Plugin.Tvdb/Providers/TvdbMissingEpisodeProvider.cs:102]: Hey, this was part of the old code when i was migrating from v3 to v4. I didn't remove it just in case if it broke anything. Think this was used when TVDb data was dirtier then current. As far as I can tell, it is no longer needed since according to the schema from TVDb, it will not be null.

**Review by crobibero — APPROVED** [review]

**Comment by crobibero** [comment]
Thank you very much for going through this adventure!

**Pull request merged**
